util/chunk/mutrow.go

```diff
@@ -0,0 +1,350 @@
+	case int64:
+		return makeMutRowUint64Column(uint64(x))
+	case uint64:
+		return makeMutRowUint64Column(x)
+	case float64:
+		return makeMutRowUint64Column(math.Float64bits(x))
+	case float32:
+		col := newMutRowFixedLenColumn(4)
+		*(*uint32)(unsafe.Pointer(&col.data[0])) = math.Float32bits(x)
+		return col
+	case string:
+		return makeMutRowBytesColumn(hack.Slice(x))
+	case []byte:
+		return makeMutRowBytesColumn(x)
+	case types.BinaryLiteral:
+		return makeMutRowBytesColumn(x)
+	case *types.MyDecimal:
+		col := newMutRowFixedLenColumn(types.MyDecimalStructSize)
+		*(*types.MyDecimal)(unsafe.Pointer(&col.data[0])) = *x
+		return col
+	case types.Time:
+		return makeMutRowInterfaceColumn(x)
+	case json.JSON:
+		return makeMutRowInterfaceColumn(x)
+	case types.Duration:
+		col := newMutRowFixedLenColumn(16)
+		*(*types.Duration)(unsafe.Pointer(&col.data[0])) = x
+		return col
+	case types.Enum:
+		col := newMutRowVarLenColumn(len(x.Name) + 8)
+		*(*uint64)(unsafe.Pointer(&col.data[0])) = x.Value
+		copy(col.data[8:], x.Name)
+		return col
+	case types.Set:
+		col := newMutRowVarLenColumn(len(x.Name) + 8)
+		*(*uint64)(unsafe.Pointer(&col.data[0])) = x.Value
+		copy(col.data[8:], x.Name)
+		return col
+	default:
+		return makeMutRowInterfaceColumn(x)
+	}
+}
+
+func newMutRowFixedLenColumn(elemSize int) *column {
+	buf := make([]byte, elemSize+1)
+	col := &column{
+		length:     1,
+		elemBuf:    buf[:elemSize],
+		data:       buf[:elemSize],
+		nullBitmap: buf[elemSize:],
+	}
+	col.nullBitmap[0] = 1
+	return col
+}
+
+func newMutRowVarLenColumn(valSize int) *column {
+	buf := make([]byte, valSize+1)
+	col := &column{
+		length:     1,
+		offsets:    []int32{0, int32(valSize)},
+		data:       buf[:valSize],
+		nullBitmap: buf[valSize:],
+	}
+	col.nullBitmap[0] = 1
+	return col
+}
+
+func makeMutRowUint64Column(val uint64) *column {
+	buf := make([]byte, 9)
+	col := &column{
+		length:     1,
+		data:       buf[:8],
+		nullBitmap: buf[8:],
+	}
+	*(*uint64)(unsafe.Pointer(&col.data[0])) = val
+	col.nullBitmap[0] = 1
+	return col
+}
+
+func makeMutRowBytesColumn(bin []byte) *column {
+	buf := make([]byte, len(bin)+1)
+	col := &column{
+		length:     1,
+		offsets:    []int32{0, int32(len(bin))},
+		data:       buf[:len(bin)],
+		nullBitmap: buf[len(bin):],
+	}
+	copy(col.data, bin)
+	col.nullBitmap[0] = 1
+	return col
+}
+
+func makeMutRowInterfaceColumn(in interface{}) *column {
+	col := &column{
+		length:     1,
+		nullBitmap: []byte{1},
+		ifaces:     []interface{}{in},
+	}
+	return col
+}
+
+// SetRow sets the MutRow with Row.
+func (mr MutRow) SetRow(row Row) {
+	for colIdx, rCol := range row.c.columns {
+		mrCol := mr.c.columns[colIdx]
+		elemLen := len(rCol.elemBuf)
+		if elemLen > 0 {
+			copy(mrCol.data, rCol.data[row.idx*elemLen:(row.idx+1)*elemLen])
+		} else if len(rCol.offsets) > 0 {
+			setMutRowBytes(mrCol, rCol.data[rCol.offsets[row.idx]:rCol.offsets[row.idx+1]])
+		} else {
+			mrCol.ifaces[0] = rCol.ifaces[row.idx]
```

Comment:
mrCol.ifaces[0] will point at the same address as rCol.ifaces[row.idx] ?

Reply:
Yes, the interface value itself is never modified.